src/ui/textfield.rs

```diff
@@ -85,56 +152,206 @@ impl TextField {
+                    0.0,
+                );
+                canvas.draw_rrect(rrect, &paint);
+            });
+        }
+
+        ui.text(canvas, &self.text_utf8, colors.text, (AlignH::Left, AlignV::Middle));
+
         canvas.restore();
         ui.pop_group();
 
-        // process events
+        // Process events
         self.process_events(ui, input);
 
         ui.pop_group();
     }
 
+    // Get selection content
+    fn selection_text(&self) -> String {
+        if self.selection.len() == 0 {
+            return String::new()
+        }
+
+        self.text[self.selection.normalize()].iter().collect()
+    }
+
+    // Set text
+    fn set_text(&mut self, text: String) {
+        self.text = text.chars().collect();
+        self.update_utf8();
+
+        self.selection.move_to(self.text.len());
+    }
+
     /// Resets the text field's blink timer.
     fn reset_blink(&mut self, input: &Input) {
         self.blink_start = input.time_in_seconds();
     }
 
-    /// Appends a character to the end of the text.
+    /// Appends a character to the cursor position.
+    /// Or replaces selection if any.
     fn append(&mut self, ch: char) {
-        self.text.push(ch);
+        if self.selection.len() > 0 {
+            self.text.splice(self.selection.normalize(), vec![ch]);
+
+            self.selection.move_to(self.selection.start() + 1);
+        } else {
+            self.text.insert(self.selection.cursor, ch);
+            self.selection.move_right(false);
+        }
+
         self.update_utf8();
     }
 
-    /// Removes a character from the end of the text.
+    /// Removes a character at cursor position.
+    /// Or removes selection if any.
     fn backspace(&mut self) {
-        self.text.pop();
+        if self.selection.len() != 0 {
+            self.delete();
+        } else if self.selection.cursor > 0 {
+            self.selection.move_left(false);
+            self.text.remove(self.selection.cursor);
+        }
+
         self.update_utf8();
     }
 
+    /// Removes character after cursor position.
+    /// Or selection if any.
+    fn delete(&mut self) {
+        if self.selection.len() != 0 {
+            self.text.drain(self.selection.normalize());
+            self.selection.move_to(self.selection.start());
+        } else if self.selection.cursor != self.text.len() {
+            self.text.remove(self.selection.cursor);
+        }
+
+        self.update_utf8();
+    }
+
+    fn key_ctrl_down(&self, input: &Input) -> bool {
+        input.key_is_down(VirtualKeyCode::LControl) || input.key_is_down(VirtualKeyCode::RControl)
+    }
+
+    fn key_shift_down(&self, input: &Input) -> bool {
+        input.key_is_down(VirtualKeyCode::LShift) || input.key_is_down(VirtualKeyCode::RShift)
+    }
+
+    fn process_word_skipping_and_selection(
+        &mut self,
+        input: &Input,
+        not_found_ws_value: usize,
+        range: Range<usize>,
+        right: bool,
+        is_shift_down: bool,
+    ) {
```

Comment:
I'm not too fond of the fact you have two `bool` parameters here. It's quite easy to confuse them. I'd hoist the `right` parameter to an `enum ArrowKey { Left, Right }`.

I'm also not sure what `not_found_ws_value`'s purpose is. Maybe you choose a more self-explanatory name, or leave a comment explaining what it is?